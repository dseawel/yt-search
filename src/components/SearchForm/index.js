import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {
  fetchYouTubeSearchResults,
  setMaxResults,
  setSearchParam,
  setSearchSort,
} from 'actions/yt-search';
import SearchInput from './SearchInput';
import SearchSort from './SearchSort';
import MaxResults from './MaxResults';
import {Button, CircularProgress} from '@material-ui/core';

export class SearchForm extends PureComponent {
  static propTypes = {
    error: PropTypes.object,
    fetchYouTubeSearchResults: PropTypes.func,
    maxResults: PropTypes.string,
    setMaxResults: PropTypes.func,
    searchParam: PropTypes.string,
    setSearchParam: PropTypes.func,
    searchSort: PropTypes.string,
    setSearchSort: PropTypes.func,
  };

  constructor(props) {
    super(props);

    this.state = {
      displayErrors: false,
      isValid: false,
      loading: false,
      validationErrors: {},
    };
  }

  _handleSearch = async () => {
    await this._validateForm();
    if (this.state.isValid) {
      this.setState({loading: true});
      this.props
        .fetchYouTubeSearchResults(this.props)
        .then(() => {
          this.setState({loading: false});
        })
        .catch(error => {
          alert('Error from YouTube');
          console.log('error', error);
        });
    }
  };

  _validateForm = () => {
    const {searchParam} = this.props;
    const validationErrors = {};
    if (!searchParam || searchParam.length < 2) {
      validationErrors.searchParam = 'Search keyword is missing';
    }
    this.setState({
      validationErrors,
      isValid: Object.keys(validationErrors).length === 0,
    });
  };

  _handlesetSearchParam = searchParam => {
    this.props.setSearchParam({searchParam});
  };

  _handlesetSearchSort = searchSort => {
    this.props.setSearchSort({searchSort});
  };

  _handlesetMaxResults = maxResults => {
    this.props.setMaxResults({maxResults});
  };

  render() {
    return (
      <div>
        <div>
          <SearchInput
            handlesetSearchParam={this._handlesetSearchParam}
            error={this.state.validationErrors.searchParam}
          />
          &nbsp;
          <MaxResults
            handlesetMaxResults={this._handlesetMaxResults}
            maxResults={this.props.maxResults}
          />
          &nbsp;
          <SearchSort
            handlesetSearchSort={this._handlesetSearchSort}
            searchSort={this.props.searchSort}
          />
        </div>
        <br />
        <div>
          <Button
            size="medium"
            color="primary"
            variant="contained"
            onClick={this._handleSearch}
          >
            {this.state.loading ? (
              <CircularProgress color="inherit" size={20} />
            ) : (
              'Search'
            )}
          </Button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ytSearch}) => {
  return {...ytSearch};
};

export default connect(
  mapStateToProps,
  {
    fetchYouTubeSearchResults,
    setMaxResults,
    setSearchParam,
    setSearchSort,
  }
)(SearchForm);
