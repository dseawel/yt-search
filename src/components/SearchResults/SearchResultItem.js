import React from 'react';
import PropTypes from 'prop-types';
import format from 'date-fns/format';
import {
  Card,
  CardHeader,
  CardMedia,
  CardContent,
  Typography,
} from '@material-ui/core';
import {withStyles} from '@material-ui/core/styles';

const propTypes = {
  classes: PropTypes.object,
  searchResultItem: PropTypes.object,
};

const styles = {
  card: {
    maxWidth: 520,
    marginBottom: 20,
  },
  media: {
    height: 0,
    paddingTop: '56.25%',
  },
};

const SearchResultItem = ({searchResultItem, classes}) => {
  const {
    snippet: {
      title,
      publishedAt,
      description,
      thumbnails: {
        high: {url: thumbnailUrl},
      },
    },
    id: {videoId},
    commentCount,
  } = searchResultItem;
  return (
    <Card className={classes.card}>
      <a
        href={`https://www.youtube.com/watch?v=${videoId}`}
        target="_blank"
        rel="noopener noreferrer"
      >
        <CardHeader
          title={title}
          subheader={format(publishedAt, 'MMM D, YYYY h:mm a')}
        />
      </a>
      <CardMedia image={thumbnailUrl} title="title" className={classes.media} />
      <CardContent>
        <Typography component="p">{description}</Typography>
        {commentCount && (
          <Typography component="em">
            {`Total Comments: ${commentCount}`}
          </Typography>
        )}
      </CardContent>
    </Card>
  );
};

SearchResultItem.propTypes = propTypes;

export default withStyles(styles)(SearchResultItem);
