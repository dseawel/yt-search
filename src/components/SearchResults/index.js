import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import SearchResultItem from './SearchResultItem';

const propTypes = {
  searchResults: PropTypes.arrayOf(PropTypes.object),
};

class SearchResults extends PureComponent {
  static propTypes = {
    searchResults: PropTypes.arrayOf(PropTypes.object),
  };

  render() {
    const {searchResults} = this.props;

    if (!searchResults) {
      return <p>{'No results to display.'}</p>;
    }

    return searchResults.map(searchResultItem => (
      <SearchResultItem
        key={searchResultItem.etag}
        searchResultItem={searchResultItem}
      />
    ));
  }
}

SearchResults.propTypes = propTypes;

const mapStateToProps = ({ytSearch: {searchResults}}) => {
  return {searchResults};
};

export default connect(
  mapStateToProps,
  {}
)(SearchResults);
