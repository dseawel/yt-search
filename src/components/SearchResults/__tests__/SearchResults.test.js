import React from 'react';
import renderer from 'react-test-renderer';
import {SearchResults} from 'components/SearchResults';

test('SearchResults renders without crashing', () => {
  const searchResults = [
    {
      commentCount: 0,
      id: {videoId: '5MI6GE0zeXg'},
      snippet: {
        publishedAt: '2019-03-25T19:12:21.000Z',
        title: 'Elite GamerEx vs JackHenry SUB TO MEH',
        thumbnails: {
          high: {
            url: 'https://i.ytimg.com/vi/5MI6GE0zeXg/hqdefault.jpg',
          },
        },
      },
      etag: '1',
    },
    {
      commentCount: 0,
      id: {videoId: 'fgtBbWTPL0w'},
      snippet: {
        publishedAt: '2012-12-22T07:22:04.000Z',
        title: 'Lumber-Jack Blues - Henry Mancini - Sometimes A Great Notion',
        thumbnails: {
          high: {
            url: 'https://i.ytimg.com/vi/fgtBbWTPL0w/hqdefault.jpg',
          },
        },
        description:
          'https://www.facebook.com/1960sFilmSoundtracksAndMovieFun Track 4 from the original soundtrack - composed by Henry Mancini from the 1970 film starring ...',
      },
      etag: '2',
    },
  ];
  const emptyComponent = <SearchResults searchResults={[]} />;
  let rendered = renderer.create(emptyComponent).toJSON();
  expect(rendered).toMatchSnapshot();
  const listComponent = <SearchResults searchResults={searchResults} />;
  rendered = renderer.create(listComponent).toJSON();
  expect(rendered).toMatchSnapshot();
});
