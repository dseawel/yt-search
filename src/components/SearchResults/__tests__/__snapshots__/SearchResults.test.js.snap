// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults renders without crashing 1`] = `null`;

exports[`SearchResults renders without crashing 2`] = `
Array [
  <div
    className="MuiPaper-root-4 MuiPaper-elevation1-7 MuiPaper-rounded-5 MuiCard-root-3 SearchResultItem-card-1"
  >
    <a
      href="https://www.youtube.com/watch?v=5MI6GE0zeXg"
      target="_blank"
    >
      <div
        className="MuiCardHeader-root-31"
      >
        <div
          className="MuiCardHeader-content-34"
        >
          <span
            className="MuiTypography-root-37 MuiTypography-headline-42 MuiCardHeader-title-35"
          >
            Elite GamerEx vs JackHenry SUB TO MEH
          </span>
          <span
            className="MuiTypography-root-37 MuiTypography-body1-46 MuiTypography-colorTextSecondary-70 MuiCardHeader-subheader-36"
          >
            Mar 25, 2019 02:12
          </span>
        </div>
      </div>
    </a>
    <div
      className="MuiCardMedia-root-73 SearchResultItem-media-2"
      style={
        Object {
          "backgroundImage": "url(\\"https://i.ytimg.com/vi/5MI6GE0zeXg/hqdefault.jpg\\")",
        }
      }
      title="title"
    />
    <div
      className="MuiCardContent-root-75"
    >
      <p
        className="MuiTypography-root-37 MuiTypography-body1-46"
      >
        undefined
      </p>
      0
    </div>
  </div>,
  <div
    className="MuiPaper-root-4 MuiPaper-elevation1-7 MuiPaper-rounded-5 MuiCard-root-3 SearchResultItem-card-1"
  >
    <a
      href="https://www.youtube.com/watch?v=fgtBbWTPL0w"
      target="_blank"
    >
      <div
        className="MuiCardHeader-root-31"
      >
        <div
          className="MuiCardHeader-content-34"
        >
          <span
            className="MuiTypography-root-37 MuiTypography-headline-42 MuiCardHeader-title-35"
          >
            Lumber-Jack Blues - Henry Mancini - Sometimes A Great Notion
          </span>
          <span
            className="MuiTypography-root-37 MuiTypography-body1-46 MuiTypography-colorTextSecondary-70 MuiCardHeader-subheader-36"
          >
            Dec 22, 2012 01:22
          </span>
        </div>
      </div>
    </a>
    <div
      className="MuiCardMedia-root-73 SearchResultItem-media-2"
      style={
        Object {
          "backgroundImage": "url(\\"https://i.ytimg.com/vi/fgtBbWTPL0w/hqdefault.jpg\\")",
        }
      }
      title="title"
    />
    <div
      className="MuiCardContent-root-75"
    >
      <p
        className="MuiTypography-root-37 MuiTypography-body1-46"
      >
        https://www.facebook.com/1960sFilmSoundtracksAndMovieFun Track 4 from the original soundtrack - composed by Henry Mancini from the 1970 film starring ...
      </p>
      0
    </div>
  </div>,
]
`;
